[B5] Migration Tooling Improvements

biyeun opened the pull request:
## Technical Summary
Some tweaks and improvements to the migration tool `migrate_app_to_bootstrap5` and cleanups in related tools.

🐟  Review commit-by-commit. Commits are very small and individually easy to digest.

highlights:
- ⭐ adds a `--no-split` option in `migrate_app_to_bootstrap5` for smaller app migration workflows, allowing migrations of templates and javascript IN PLACE (no file splitting). The command then pauses after the migration changes are made and prompts you to add the `@use_bootstrap5` decorator to the relevant views before moving to the next file.
- ⭐ Updates the `--template-name` and `--js-name` options so that if a `bootstrap3` path is specified, it will re-migrate that `bootstrap3` file and then update the associated `bootstrap5` file, allowing quick resolutions of situations where other folks have made changes to the split templates while the split files PR was in progress. It also allows for re-migrating files that were previously migrated to apply new automation/improvements from `migrate_app_to_bootstrap5` if it's been a while since that file has been migrated.
- 📄 Docs have been updated for the above options
- 🐞 fixes the issue in `complete_bootstrap5_migration` where `case_search/case_search.html` got stored as `.html` in the completed file list 🤦‍♀️ 
- 🧹  cleanup / tidying of small things here and there.
- 🚫 removes the `--re-check` option from `migrate_app_to_bootstrap5` as the changes to `--template-name` and `--js-name` are much better solutions for the original intention of that option. I don't think we want to apply migrations again to `bootstrap5` templates ever, as that will not be valid for things like `col-<size>` updates, and will involve actually having to go through things change-by-change.


## Safety Assurance

### Safety story
Only touches internal tools. Important parts have tests. :)

### Automated test coverage
Yes, the important bits are tested

### QA Plan
Not needed

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

biyeun committed:
use removeprefix instead of lstrip to fix issues like 'case_search/case_search.html' > '.html'

biyeun committed:
add ability to auto-commit changes

biyeun committed:
add advice for where the logs dir should be located

biyeun committed:
add ability to skip the split files step and migrate files in place (for smaller apps)

biyeun committed:
provide default for apply changes

biyeun committed:
remove extra parenthesis that snuck in there

biyeun committed:
make sure skip all cannot be used with template-name and js-name options

biyeun committed:
remove un-useful re-check option and replace with workflow to re-migrate a bootstrap3 template AFTER it has been split

biyeun committed:
if no changes were necessary, indicate that to the user

biyeun committed:
update documentation for new options in migrate_app_to_bootstrap5

biyeun committed:
only show next steps for templates

biyeun committed:
add nudge after next steps during migrating app in place

biyeun committed:
don't hard block user from continuing, as some folks don't like this approach

biyeun committed:
suggest commit message for marking app as complete with option to apply

biyeun committed:
pause for user to commit changes before proceeing

biyeun committed:
make enter_to_continue its own utility

orangejenny commented on corehq/apps/hqwebapp/management/commands/migrate_app_to_bootstrap5.py:
```diff
@@ -127,18 +134,35 @@ def handle(self, app_name, **options):
 
+        self.no_split = options.get('no_split')
         self.skip_all = options.get('skip_all')
+        if self.skip_all and self.no_split:
```
Curious why this is? You can review the changes in the code, especially if you're not auto-committing.

biyeun replied:
It's because we prompt the user to update the relevant views with the `@use_bootstrap5` decorator after each file is migrated in-place.

biyeun replied:
and `--skip-all` by default auto-commits everything

orangejenny replied:
ah, I blanked on what all `skip-all` does

orangejenny commented on corehq/apps/hqwebapp/management/commands/migrate_app_to_bootstrap5.py:
```diff
@@ -69,6 +70,12 @@ def add_arguments(self, parser):
             help="Run migration against already split bootstrap 5 files"
         )
+        parser.add_argument(
+            '--no-split',
```
Thanks!

orangejenny commented on corehq/apps/hqwebapp/management/commands/migrate_app_to_bootstrap5.py:
```diff
@@ -142,6 +142,12 @@ def handle(self, app_name, **options):
             )
             return
 
+        if self.skip_all and (js_name or template_name):
```
Why is this?

biyeun replied:
because we want the user to confirm the correct file paths were picked up. often this is only one file anyway, so `--skip-all` doesn't make much of a difference. Additionally, we want to prompt the user differently if the file is a `bootstrap3` re-migration.

orangejenny commented on corehq/apps/hqwebapp/management/commands/migrate_app_to_bootstrap5.py:
```diff
@@ -439,6 +419,33 @@ def show_next_steps_after_migrating_file_in_place(self, short_path):
         )
         self.enter_to_continue()
 
+    def migrate_file_again(self, app_name, bootstrap3_path, bootstrap5_lines, is_template):
```
I don't quite understand this workflow, but defer to @jingcheng16 if she does. I had thought this would re-migrate the B5 file. Doesn't re-migrating the B3 file mean that any changes made to the B5 file, like fixing flagged issues flagged in the original migration, will be overwritten?

biyeun replied:
yes. But those flagged issues are often not fixed as part of the split files PR workflow for large apps, only the automated changes are committed in those cases. In that case, the only changes in the `bootstrap5` templates are the auto-migrated changes that this command is just redoing anyway.

orangejenny commented on corehq/apps/styleguide/templates/styleguide/bootstrap5/migration_guide.html:
```diff
@@ -144,44 +160,45 @@ <h2 id="split-files" class="pt-4">
-    After applying all the actions related to splitting a file, it is recommended to commit those changes before
-    moving on to the next file. The management command will prompt you to do this with a suggested commit message.
-    Formatting our commits this way is important to maintain a clean git history.
+    Regardless of what route you take, a changelog for each file will be saved to a markdown file in
```
Idle thought: I would *love* to write one changelog for automated fixes and one for flagged issues (and will probably make that update eventually). I start every view migration by editing the change log, which is hundreds of lines long even for the small migrations I've been doing, and I either delete all the automated changes and re-save the log, or I make TODO comments in the code for the flagged issues and then work off of those and forget about the log file.

biyeun replied:
ah, i do like this thought. how about adding `#todo` comments around flagged areas as well? I think that would be helpful too

biyeun replied:
would you like me to include these improvements in this PR or a followup?

orangejenny replied:
TODO comments would be amazing. Definitely doesn't need to be part of this PR.

orangejenny commented on corehq/apps/hqwebapp/management/commands/migrate_app_to_bootstrap5.py:
```diff
@@ -124,32 +129,54 @@ def handle(self, app_name, **options):
             ))
             ensure_no_pending_changes_before_continuing()
 
+        if self.no_split:
```
I don't think this is necessary. This isn't a dangerous or un-revert-able operation.

biyeun replied:
removed, as well as the prompt for `self.skip_all`

orangejenny commented on corehq/apps/styleguide/templates/styleguide/bootstrap5/migration_guide.html:
```diff
@@ -265,6 +282,72 @@ <h3 id="verify-references" class="pt-3">
+    please run the following command to re-migrate the <code>bootstrap3</code> version of that file.
+  </p>
+  <p>
+    If the file is a template, please run:
```
It'd be a nice simplification to have a single `--file-name` argument and have the tool detect whether it's a js or html file.

biyeun replied:
do you want me to include this in this PR or do a followup PR?

orangejenny replied:
either, doesn't need to be a priority

biyeun replied:
ended up including the change :)

orangejenny commented on corehq/apps/hqwebapp/management/commands/complete_bootstrap5_migration.py:
```diff
@@ -46,7 +45,7 @@ def handle(self, app_name, **options):
             self.stdout.write(self.style.ERROR(
                 "You have un-committed changes. Please commit these changes before proceeding...\n"
             ))
-            ensure_no_pending_changes_before_continuing()
```
Thank you. Deleted the screed I wrote on the commit that introduced this.

biyeun committed:
merge --template-name and --js-name to just be --filename

biyeun committed:
allow skip all and filename to be used

biyeun committed:
remove confirmation prompt for skip all and no split options

orangejenny commented on corehq/apps/hqwebapp/management/commands/migrate_app_to_bootstrap5.py:
```diff
@@ -131,27 +131,12 @@ def handle(self, app_name, **options):
             )
             return
 
-        if self.skip_all:
```
Lovely. 🤠

jingcheng16 commented on corehq/apps/hqwebapp/utils/bootstrap/status.py:
```diff
@@ -71,7 +71,7 @@ def mark_javascript_as_complete(app_name, js_short_path):
     app_summary = get_app_status_summary(app_name)
     app_summary.setdefault(file_type, []).append(
-        short_path.lstrip(f"{app_name}/")
+        short_path.removeprefix(f"{app_name}/")
```
Hi Biyeun, why `removeprefix` will help in this situation? (Just be curious and I cannot figure it out myself by reading the doc of `lstrip` and `removeprefix`)

biyeun replied:
`removeprefix` only removes the string one time, while `lstrip` removes multiple. you can also test it in shell

jingcheng16 commented on corehq/apps/hqwebapp/management/commands/migrate_app_to_bootstrap5.py:
```diff
@@ -86,6 +86,9 @@ def handle(self, app_name, **options):
             self.stderr.write("\nPlease make sure BOOTSTRAP_MIGRATION_LOGS_DIR is "
                               "set in your localsettings.py before continuing...\n\n")
+            self.stdout.write(self.style.MIGRATE_LABEL(
+                "TIP: path should be outside of this repository\n\n"
```
lol... thank you!!!

jingcheng16 reviewed (approved):
Looks good!! Thank you for those improvements!

biyeun committed:
make is_split_path a utility

biyeun committed:
--filename can use bootstrap5 files, with restrictions
allow users to specify a bootstrap5 path to re-migrate. However, do not allow certain migrations to be re-run (like renaming column names)

biyeun committed:
update migrate_file_again to handle bootstrap5 files

biyeun commented:
@jingcheng16 hopping online briefly this morning since my night yesterday went to spending time with fam :) I added some changes relevant to @orangejenny's PR here https://github.com/dimagi/commcare-hq/pull/34498#issuecomment-2077263786 which allows re-migrating a `bootstrap5` file with restrictions :)

biyeun committed:
updates to handling --filname option
make sure that if a bootstrap5 path was not explicitly specified, and files returns multiple results, we filter out all the bootstrap5 paths. otherwise, we risk migrating both the bootstrap5 and bootstrap3 version of the same file, which does not make sense. This is the case when a user only specifies the filename, rather than the relative path

biyeun committed:
update javascript dependency regex to handle hqDefine lines, with test
